### PR TITLE
Use storage.local instead of localStorage

### LINKED ISSUE #3076: Make our background storage more size resilient
Related to #3071

Meta-issue to address #3071.

We now know that storage is overflowing for some users, and this may be causing vault loss.

We need to develop strategies to improve our storage size resilience, and implement them.

- [x] Study what happens when background localstorage cannot write anymore. Is the storage instantly lost? Is there an opportunity to backup/recover?
- [x] Study which portions of our persisted state take the most size (probably transactions).
- [x] Review extension storage limits.
- [x] Research other alternative storage strategies.
- [ ] Move our `localStorage` use to the better solution (`storage.local`).

### PR DESCRIPTION
**Pull request by danfinlay**
Fixes #3076, needs thorough QA:

- Create a vault on `master`, rebuild on this branch, see if MM continues seamlessly.
- Try destroying the old `localStorage` data manually, see if MM continues seamlessly.
- Do [full QA](https://consensys.quip.com/YmJuANWLkEAv) on this branch, to ensure it's internally reliable.
- Maybe try filling up storage manually somehow. Could be by creating tons of accounts or tons of testnet transactions that publish large contracts.

### CODE REVIEW

File: app/scripts/background.js [hunk 1]
```diff
@@ -100,6 +117,17 @@ function setupController (initState) {
     return versionedData
   }
 
+  function syncDataWithExtension(state) {
+    if (localStore.isSupported) {
+      try {
+        localStore.set(state)
+      } catch (err) {
+        log.error('error setting state in local store:', err)
```
Comment (kumavis) [inline app/scripts/background.js]: i dont think these get submitted to sentry (but maybe they do)

File: app/scripts/lib/extension-store.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+module.exports = class ExtensionStore {
+  constructor() {
+    this.isSupported = !!(extension.storage && extension.storage.sync)
+    this.isEnabled = true // TODO: get value from user settings
+  }
+  async fetch() {
+    return new Promise((resolve) => {
+      extension.storage && extension.storage.sync.get(KEYS_TO_SYNC, (data) => {
```
Comment (kumavis) [inline app/scripts/lib/extension-store.js]: we wanted to drop `KEYS_TO_SYNC`, right?

File: app/scripts/background.js [hunk 3]
```diff
@@ -52,6 +54,20 @@ async function loadStateFromPersistence () {
...
+
+  if (localStore.isSupported) {
+    let localData
+    try {
+      localData = await localStore.get()
+    } catch (err) {
+      log.error('error fetching state from local store:', err)
+    }
+
+   // TODO: handle possible exceptions (https://developer.chrome.com/apps/runtime#property-lastError)
```
Comment (kumavis) [inline app/scripts/background.js]: looks like this is somewhat handled in `ExtensionStore`
```js
 const lastError = extension.runtime.lastError
```

File: app/scripts/lib/local-store.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+const STORAGE_KEY = 'metamask-config'
+
+module.exports = class ExtensionStore {
+  constructor() {
+    this.isSupported = !!(extension.storage.local)
+    if (!this.isSupported) {
+      log.error('Storage local API not available.')
+    }
+  }
+  async get() {
```
Comment (kumavis) [inline app/scripts/lib/local-store.js]: doesnt need to be marked as `async` when manually returning a promise but maybe this is a good practice anyways

**Review by kumavis — COMMENTED** [review]
Nice -- looking forward to it

added some questions / comments

File: app/scripts/background.js (hunk 1 above)
Comment (danfinlay) [reply app/scripts/background.js]: What does get submitted for sure?

Comment (danfinlay) [reply app/scripts/background.js]: (we should document our correct logging practice)

File: app/scripts/lib/extension-store.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+const extension = require('extensionizer')
+
+const KEYS_TO_SYNC = ['KeyringController', 'PreferencesController']
+const FIREFOX_SYNC_DISABLED_MESSAGE = 'Please set webextensions.storage.sync.enabled to true in about:config'
+
+const handleDisabledSyncAndResolve = (resolve, toResolve) => {
```
[the remark this answers is not in the document]
Comment (danfinlay) [reply app/scripts/lib/extension-store.js]: This is part of the original Ellieday PR, we aren't actually using this file in this PR, so I'm not sure I want to bother cleaning it up, I'd rather either:
- Delete the file and its test
- Ignore it for now

File: app/manifest.json [hunk 6]
```diff
@@ -60,6 +60,7 @@
     "http://localhost:8545/",
     "https://*.infura.io/"
     ],
+  "unlimitedStorage": true,
```
Comment (tmashuang) [inline app/manifest.json]: `unlimitedStorage` should be added to the permissions, the same as `storage`. Referenced: [Declare Permissions](https://developer.chrome.com/extensions/declare_permissions)

**Comment by kumavis** [comment]
@heyellieday we used some of your code in this PR so we'd like to credit you with partial bounty reward of the original bounty (synced storage). Can you give an ethereum address here or emailed to `aaron at kumavis dot me` thanks/

File: app/scripts/background.js (hunk 3 above)
Comment (kumavis) [reply app/scripts/background.js]: I'm hesitant to merge a TODO in here, especially around error handling

File: app/scripts/background.js (hunk 1 above)
Comment (kumavis) [reply app/scripts/background.js]: im not certain atm, but you can test it with the `Test-Metamask` sentry project on dev

**Review by kumavis — APPROVED** [review]

**Comment by kumavis** [comment]
cool, just needs QA then

**Comment by danfinlay** [comment]
Added to that branch:
- No longer writes to `localStorage`.
- Fixed issue where quantity of write attempts overwhelmed our `pump` to `storage.local` stream.

**Comment by heyellieday** [comment]
@kumavis, sorry, I missed these mentions!

I'm glad some of my work was able to be used for this issue and thanks for crediting me part of the bounty! Here's an ethereum address to send to: 0x1ab4C79D96a05b64742AE1B0213B84fB984917C8.

I'm hoping I can help out with more in the future :)

File: app/scripts/background.js [hunk 7]
```diff
@@ -65,11 +69,34 @@ async function loadStateFromPersistence () {
...
+  if (localStore.isSupported) {
+    let localData
+    try {
+      localData = await localStore.get()
+    } catch (err) {
+      log.error('error fetching state from local store:', err)
+    }
+
+    if (Object.keys(localData).length > 0) {
+      versionedData = localData
```
Comment (kumavis) [inline app/scripts/background.js]: why do we do this
please add a comment

File: app/scripts/background.js [hunk 8]
```diff
@@ -65,11 +69,34 @@ async function loadStateFromPersistence () {
   // migrations
   const migrator = new Migrator({ migrations })
   // read from disk
-  let versionedData = diskStore.getState() || migrator.generateInitialState(firstTimeState)
+  versionedData = diskStore.getState() || migrator.generateInitialState(firstTimeState)
```
Comment (kumavis) [inline app/scripts/background.js]: maybe something like
`versionedData = localData || diskStore.getState() || migrator.generateInitialState(firstTimeState)`

Comment (danfinlay) [reply app/scripts/background.js]: New commit moves storage.local loading before this assignment so I can use a similar format to what you suggest.

File: app/scripts/background.js (hunk 7 above)
Comment (danfinlay) [reply app/scripts/background.js]: Comment added.

**Review by kumavis — DISMISSED** [review]
im going to take another pass over this, please wait

**Review by kumavis — APPROVED** [review]

**Comment by lazaridiscom** [comment]
@danfinlay , what about the old installations, which use the old storage? Are they migrated to the new storage model?

**Comment by danfinlay** [comment]
Yes, they are migrated to the new storage model, and the old storage model is kept around, in case the new one fails.

**Pull request merged**